### PR TITLE
Add Google OAuth support

### LINKED ISSUE #379: Develop: Implement Google OAuth email retrieval
As a user of the application, I want to be able to log in with any common email service.

Acceptance Criteria
--
- User has the option to log in to their Gmail account using OAuth
- User has the option to log in with the current username/password approach
- Document any relevant content in the Programmer guide (draft thoughts of usage and importantly the "Why")
- Updates SRS and STP (and other docs if necessary)
- Add/Update relevant unit tests

Example of Apple's Mail login:
Image

### PR DESCRIPTION
**Pull request by JesterSe7en**
- Add Google OAuth sign in - use read only scope
- Rename AuthService to UserAuthService
- Use enough_mail MediaType parser to check Mimedata
- Sign out of google when tapping on exit
- If user is using google sign in, parse emails differently

### CODE REVIEW

**Comment by shaneknows** [comment]
I'm sure it's very close, but I'm still getting Login Credentials Failed. Is there anything i need to do to get it working locally/with downloaded apk?

image

File: lib/utility/user_auth_service.dart [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    try {
+      await _googleSignIn.signIn();
+      return _googleSignIn.isSignedIn();
+    } catch (error) {
+      print(error);
+      return false;
+    }
+  }
+
+  // Future<AuthClient> getGoogleAuthenticatedClient() async {
```
Comment (Minyahil40) [inline lib/utility/user_auth_service.dart]: you can remove unused methods

File: lib/ui/sign_in.dart [hunk 2]
```diff
@@ -574,18 +592,85 @@ class SignInWidgetState extends AssistantState<SignInWidget> {
...
+                                    text: 'Sign In with Google',
+                                  ),
+                                ))
+                              ],
+                            ),
+                            Container(
+                              padding: const EdgeInsets.only(top: 10),
+                            ),
+                            Row(
+                              children: [
```
Comment (Minyahil40) [inline lib/ui/sign_in.dart]: apple and MS sign in require developer account to configure. you should remove it unless its work in progress.

**Review by mdconatser — DISMISSED** [review]

File: lib/ui/sign_in.dart (hunk 2 above)
Comment (JesterSe7en) [reply lib/ui/sign_in.dart]: @mdconatser Are the other sign in options being actively worked on?

**Comment by JesterSe7en** [comment]
@shaneknows  There's probably something I need to do in the developer console.

I had to generate a SHA1 hash with my debug.keystore -> upload that to Google Firebase SWEN670 project settings to get google sign-in's pop up.

If you have any knowledge on how to get this to work for everyone, let me know.  I'll continue to look up how.

**Comment by Minyahil40** [comment]
> @shaneknows There's probably something I need to do in the developer console.
> 
> I had to generate a SHA1 hash with my debug.keystore -> upload that to Google Firebase SWEN670 project settings to get google sign-in's pop up.
> 
> If you have any knowledge on how to get this to work for everyone, let me know. I'll continue to look up how.

you can generate the SHA1 key by running the following command from the project folder
For Mac:
 keytool -list -v -keystore ~/.android/debug.keystore -alias androiddebugkey -storepass android -keypass android

For Windows:
 keytool -list -v -keystore "\.android\debug.keystore" -alias androiddebugkey -storepass android -keypass android

**Comment by JesterSe7en** [comment]
> @shaneknows There's probably something I need to do in the developer console.
> 
> I had to generate a SHA1 hash with my debug.keystore -> upload that to Google Firebase SWEN670 project settings to get google sign-in's pop up.
> 
> If you have any knowledge on how to get this to work for everyone, let me know. I'll continue to look up how.

Would using the keytool command on /android/app/upload-keystore-fall2022.jks and uploading that SHA1 to firebase work?

Running this keytool + default password against that file doesn't work. says incorrect password for me.

File: lib/ui/sign_in.dart (hunk 2 above)
Comment (mdconatser) [reply lib/ui/sign_in.dart]: i don't believe anyone is working on the other ones yet

**Review by mdconatser — DISMISSED** [review]

**Comment by JesterSe7en** [comment]
https://stackoverflow.com/questions/27838636/what-is-the-use-of-debug-keystore-in-android

References using a shared debug keystore in teams to facilitate an easier time to debug app.

File: lib/ui/sign_in.dart (hunk 2 above)
Comment (JesterSe7en) [reply lib/ui/sign_in.dart]: removed apple/ms

**Review by mdconatser — DISMISSED** [review]

**Comment by shaneknows** [comment]
> https://stackoverflow.com/questions/27838636/what-is-the-use-of-debug-keystore-in-android
> 
> References using a shared debug keystore in teams to facilitate an easier time to debug app.

Ok. Sounds like if I use the version from the play store this should work. Need to fix the merge conflicts and we can get this in

**Review by Minyahil40 — DISMISSED** [review]

**Comment by mdconatser** [comment]
need to resolve conflicts

**Review by mdconatser — APPROVED** [review]

**Review by shaneknows — APPROVED** [review]

**Pull request merged**